docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,264 @@
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binaries must be packaged and distributed.
+
+The author recommends *1.*. As mentioned, we have a hard requirement
+of not allowing the core commands to be overrided. In *2.* this is
+_always_ possible.
+
+=== `radicle-bins`
+
+Per RFC-682#CLI, all subcommands will be provided as
+libraries. However, this leaves the question of where the binary might
+live. So far, it has been an unspoken formalism that all binaries live
+in `radicle-bins` where we can provide a `Cargo.lock` file. All
+functionality will live in `radicle-link`, including the `main`
+function, and the entry in `radicle-bins` will be a thin wrapper
+around this function.
+
+== Core
+
+With all the details in place, we define a non-exhaustive list of the
+core commands that we wish to include for the first iteration. More
+components MAY be added as the result of future RFCs. We SHALL keep a
+list of the core commands as part of the `radicle-link` documentation.
+
+=== `rad-project`
+
+This command will provide management of the `Project` identities and
+their respective working copies. The following actions will be
+defined:
+
+* `create` — create a `Project` identity and either create a
+  respective working copy, or link it to an existing working copy.
+* `get` — attempt to retrieve a `Project` identity from the store based on
+  a provided URN.
+* `list` — list all `Project` identities in the store.
+* `update` — update a `Project` identity with a new payload.
+* `merge` — merge a peer's version of the `Project` identity with your own
+* `track` — track a peer for the `Project` identity and add the peer
+to the working copy's include file.
+
+=== `rad-person`
+
+This command will provide the management of the `Person`
+identities. The following actions will be defined:
+
+* `create`— create a `Person` identity.
+* `get`— attempt to retrieve a `Person` identity from the store based
+  on a provided URN.
+* `list`— list all `Person` identities in the store.
+* `update`— update a `Person` identity with a new payload.
+* `merge`— merge a peer's version of the `Person` identity with your own.
+* `track`— track a peer for the `Person` identity.
+
+=== `rad-replicate`
+
+* `ask` — ask the network for a given URN and replicate it to local storage.
+
+=== `rad-key`
+
+This command will provide the management of keys and the `Profile`
+construct in `librad`. The following actions will be defined:
+
+* `create`— generate a new key and a new `Profile`, which implies a
+  separate storage.
+* `profile get`— get the current `Profile`'s identifier
```

Comment:
Are these sub-sub-commands? Or is this like, `rad profile-get`? I wonder if for plumbing commands we should keep it simple with only one level of sub-commands, like git, eg. `git branch -D` to delete a branch, not `git branch delete`

Reply:
What's the advantage of one over the other?

Reply:
I guess the git-style namespacing has these advantages:

* It's clear what's a sub-command vs just a parameter to a command, so it may lead to less ambiguities, eg. `git push origin master` is `git-push origin master` and not `git-push-origin master` or `git-push-origin-master`, especially for commands that accept positional arguments.
* You don't need to handle cases like `rad profile --flag get` vs. `rad profile get --flag`

So basically it's simpler from an implementation perspective to only have 1 level of sub-commands. The advantages of having multiple levels are:

* For CLIs with hundreds of sub-commands and a clear hierarchy, eg. `gcloud`, it may be easier to namespace everything
* It "reads" better maybe or has the promise of being more user-friendly, though I'd say that is subjective

Reply:
I've leaned towards the hierarchy of commands. I feel that subcommands are easier on the cognitive load rather than trying to remember what shorts were given for a command. I was also thinking that tab completion would work better for subcommands, which isn't possible with flags (or at least not that I know of).